internal/http/services/owncloud/ocs/handlers/apps/sharing/shares/shares.go

```diff
@@ -282,7 +282,31 @@ func (h *Handler) CreateShare(w http.ResponseWriter, r *http.Request) {
 	reqRole, reqPermissions := r.FormValue("role"), r.FormValue("permissions")
 	switch shareType {
 	case int(conversions.ShareTypeUser), int(conversions.ShareTypeGroup):
-		// user collaborations default to Manager (=all permissions)
+		// NOTE: clients tend to send "31" as permissions but they mean "15".
+		// This is because it adds the "16" for sharing , but that is now no longer allowed.
+		// We could now have some fancy mechanism that casts the string to an int, subtracts 16 and casts it back to a string.
+		// Or we could change the role later and hope everything works out.
+		// Or:
+		if reqRole == "" {
+			switch reqPermissions {
```

Comment:
The permission bits are:
https://doc.owncloud.com/server/next/admin_manual/configuration/files/file_sharing_configuration.html#permissions-masks

1 - read
2 - update
4 - create
8 - delete
16 - share (no longer used on ocis)

I suppose that someone could be granted 1+8+16=25 (read, delete and share) and that should become 9.

And for "file drop" there can be combinations like update+create=6 (the person doing the "file drop" cannot read back the file after "dropping" it), and maybe some client might send 22 as the permissions?

Why not just "cast to int" then take the remainder mod16, and cast back to string. That avoids thinking about which permission bit combinations will actually happen.

Reply:
This is a deprecated endpoint. Additionally I hope that his gets fixed in clients so we can remove it here. It doesn't need to be very resilient, only needs to fix the actual problem. In fact, it would already be enough to only handle only "17", "25" and "31". I just added the other values so it matches with the tests. I don't think we need any more values as there will not be any feature updates to this endpoint

Reply:
FYI, @TheOneRing, ok for you?